src/analyzer/analyzerqueue.cpp

```diff
@@ -35,160 +27,202 @@ mixxx::Logger kLogger("AnalyzerQueue");
 // We need to use a smaller block size, because on Linux the AnalyzerQueue
 // can starve the CPU of its resources, resulting in xruns. A block size
 // of 4096 frames per block seems to do fine.
-const mixxx::AudioSignal::ChannelCount kAnalysisChannels(mixxx::kEngineChannelCount);
-const SINT kAnalysisFramesPerBlock = 4096;
+constexpr mixxx::AudioSignal::ChannelCount kAnalysisChannels = mixxx::kEngineChannelCount;
+constexpr SINT kAnalysisFramesPerBlock = 4096;
 const SINT kAnalysisSamplesPerBlock =
         kAnalysisFramesPerBlock * kAnalysisChannels;
 
+// Maximum frequency of progress updates
+constexpr int kProgressUpdateInhibitMillis = 100;
+
+constexpr int kProgressStateEmpty   = 0;
+constexpr int kProgressStateWriting = 1;
+constexpr int kProgressStateReady   = 2;
+constexpr int kProgressStateReading = 3;
+
 QAtomicInt s_instanceCounter(0);
 
 } // anonymous namespace
 
+AnalyzerQueue::ThreadProgress::ThreadProgress()
+    : m_state(kProgressStateEmpty),
+      m_currentTrackProgress(kAnalysisProgressNone),
+      m_queueSize(0) {
+}
+
+bool AnalyzerQueue::ThreadProgress::tryWrite(
+        TracksWithProgress* previousTracksWithProgress,
+        TrackPointer currentTrack,
+        int currentTrackProgress,
+        int queueSize) {
+    DEBUG_ASSERT(previousTracksWithProgress);
+    bool wasEmpty = m_state.testAndSetAcquire(
+            kProgressStateEmpty, kProgressStateWriting);
+    bool progressChanged = false;
+    if (wasEmpty || m_state.testAndSetAcquire(
+            kProgressStateReady, kProgressStateWriting)) {
+        DEBUG_ASSERT(m_state == kProgressStateWriting);
+        //kLogger.trace() << "Writing progress info";
+        // Keep all track references alive until the main thread releases
+        // them by moving them into the progress info exchange object!
+        if (!previousTracksWithProgress->empty()) {
+            if (m_tracksWithProgress.empty()) {
+                previousTracksWithProgress->swap(m_tracksWithProgress);
+            } else {
+                for (const auto trackProgress: *previousTracksWithProgress) {
+                    m_tracksWithProgress[trackProgress.first] = trackProgress.second;
+                }
+                previousTracksWithProgress->insert(
+                        m_tracksWithProgress.begin(),
+                        m_tracksWithProgress.end());
+            }
+            previousTracksWithProgress->clear();
```

Comment:
not sure if I have understood completely this part, why does it do an insert to previousTracksWithProgress just to make a clear afterwards?

Reply:
One of the statements must have slipped through from a previous version. Of course, that doesn't make sense!